sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -133,82 +140,84 @@ object NestedColumnAliasing {
-   *
-   * 1. ExtractValue -> Alias: A new alias is created for each nested field.
-   * 2. ExprId -> Seq[Alias]: A reference attribute has multiple aliases pointing it.
+   * Return [[GetStructField]] or [[GetArrayStructFields]] on top of other [[ExtractValue]]s
+   * or special expressions.
    */
-  def getAliasSubMap(exprList: Seq[Expression], exclusiveAttrs: Seq[Attribute] = Seq.empty)
-    : Option[(Map[ExtractValue, Alias], Map[ExprId, Seq[Alias]])] = {
-    val (nestedFieldReferences, otherRootReferences) =
-      exprList.flatMap(collectRootReferenceAndExtractValue).partition {
-        case _: ExtractValue => true
-        case _ => false
-      }
+  private def collectExtractValue(e: Expression): Seq[ExtractValue] = e match {
+    case g if isSelectedField(g) => Seq(g.asInstanceOf[ExtractValue])
+    case es if es.children.nonEmpty => es.children.flatMap(collectExtractValue)
+    case _ => Seq.empty
+  }
 
-    // Note that when we group by extractors with their references, we should remove
-    // cosmetic variations.
+  /**
+   * Creates a map from root [[Attribute]]s to non-redundant nested [[ExtractValue]]s in the
+   * case that only a subset of the nested fields are used.
+   * Nested field accessors of `exclusiveAttrs` are not considered in nested fields aliasing.
+   */
+  def getAttributeToExtractValues(
+      exprList: Seq[Expression],
+      exclusiveAttrs: Seq[Attribute]): Option[Map[Attribute, Seq[ExtractValue]]] = {
+
+    val nestedFieldReferences = exprList.flatMap(collectExtractValue)
+    val otherRootReferences = exprList.flatMap(collectAttributeReference)
     val exclusiveAttrSet = AttributeSet(exclusiveAttrs ++ otherRootReferences)
-    val aliasSub = nestedFieldReferences.asInstanceOf[Seq[ExtractValue]]
+
+    // Remove cosmetic variations when we group extractors by their references
+    val attributeToExtractValues = nestedFieldReferences
       .filter(!_.references.subsetOf(exclusiveAttrSet))
       .groupBy(_.references.head.canonicalized.asInstanceOf[Attribute])
-      .flatMap { case (attr, nestedFields: Seq[ExtractValue]) =>
-        // Remove redundant `ExtractValue`s if they share the same parent nest field.
+      .flatMap { case (attr: Attribute, nestedFields: Seq[ExtractValue]) =>
+        // Remove redundant [[ExtractValue]]s if they share the same parent nest field.
         // For example, when `a.b` and `a.b.c` are in project list, we only need to alias `a.b`.
-        // We only need to deal with two `ExtractValue`: `GetArrayStructFields` and
-        // `GetStructField`. Please refer to the method `collectRootReferenceAndExtractValue`.
         val dedupNestedFields = nestedFields.filter {
-          case e @ (_: GetStructField | _: GetArrayStructFields) =>
-            val child = e.children.head
+          // See [[collectExtractValue]]: we only need to deal with [[GetArrayStructFields]] and
+          // [[GetStructField]]
+          case GetStructField(child, _, _) =>
```

Comment:
This seems to be the same as the original logic?

Reply:
This splits `collectRootReferenceAndExtractValue` into two lists, `collectExtractValue` and `collectRootReferences`. Previously, we created a list with both types, performed a partition, and then casted the `ExtractValue`s.

Reply:
Ah I meant this
```scala
case e @ (_: GetStructField | _: GetArrayStructFields) =>
  val child = e.children.head
  nestedFields.forall(f => child.find(_.semanticEquals(f)).isEmpty)
```
to this
```scala
case GetStructField(child, _, _) =>
  nestedFields.forall(f => child.find(_.semanticEquals(f)).isEmpty)
case GetArrayStructFields(child, _, _, _, _) =>
  nestedFields.forall(f => child.find(_.semanticEquals(f)).isEmpty)
```

Reply:
Ah, I just found `e.children.head` confusing. I'll revert this, I don't have that strong of an opinion.